<div style = "text-align: right"> Roger Lee </div>

<left>FINM 32000</left> 
<br>
<left>Spring 2024</left>

<h2><center>Homework # 3</center></h2>
<h3><center><font color = "blue">Solution By: Imtiaz Ali</font></center></h3>

In [1]:
import numpy as np
import pandas as pd
from scipy.stats import norm
from typing import Tuple


In [2]:
%%time
####================================================
import sys

### Get Python and Jupyter Notebook version and information
# print('=' * 50)
# print(sys.executable)
print('=' * 50)
print(sys.version)
print('=' * 50)
print(sys.version_info)
print('=' * 50)
!python --version
print('=' * 50)
!jupyter --version
print('=' * 50)


3.8.12 (default, Oct 12 2021, 03:01:40) [MSC v.1916 64 bit (AMD64)]
sys.version_info(major=3, minor=8, micro=12, releaselevel='final', serial=0)
Python 3.8.12
Selected Jupyter core packages...
IPython          : 8.12.0
ipykernel        : 6.19.2
ipywidgets       : 8.0.4
jupyter_client   : 8.1.0
jupyter_core     : 5.3.0
jupyter_server   : 1.13.5
jupyterlab       : 3.3.2
nbclient         : 0.5.13
nbconvert        : 6.1.0
nbformat         : 5.7.0
notebook         : 6.5.4
qtconsole        : 5.4.2
traitlets        : 5.7.1
CPU times: total: 31.2 ms
Wall time: 1.95 s


# Problem 1


> Assume that the short rate (the instantaneous spot rate of interest) follows the process:
$$\text{d}r_{t} = \alpha(r_{t}, t)\text{d}t + \beta(r_{t}, t)\text{d}W_{t}$$
where $W_{t}$ is Brownian motion under risk-neutral probabilities. This framework includes models such as the Vasicek and CIR models, which correspond to particular choices of the functions $\left(\alpha, \beta\right)$.


## Problem 1a


Considering an interest rate derivative with:
>- Time-$T$ terminal payout given by $F(r_{T})$
>- Time-$t$ price given by $C_{t}=C(r_{t}, t)$, where $C$ is a smooth pricing function $C$

Then we can use Ito's rule:
$$\text{d}f(X_{t}, t)=\frac{\partial f}{\partial t}\text{d}t + \frac{\partial f}{\partial x}\text{d}X_{t} + \frac{1}{2}\frac{\partial^{2} f}{\partial x^{2}}(\text{d}X_{t})^{2}$$

where we can substitute $f=C$ and $X_{t}=r_{t}$ to get:
$$\begin{align*}
\text{d}C(r_{t},t)=&\hspace{1mm}\frac{\partial C}{\partial t}\text{d}t + \frac{\partial C}{\partial r}\text{d}r_{t} + \frac{1}{2}\frac{\partial^{2} C}{\partial r^{2}}(\text{d}r_{t})^{2}\\
=&\hspace{1mm} \frac{\partial C}{\partial t}\text{d}t
 + \frac{\partial C}{\partial r}\bigg(\alpha(r_{t}, t)\text{d}t + \beta(r_{t}, t)\text{d}W_{t} \bigg)
 + \frac{1}{2} \frac{\partial^{2} C}{\partial r^{2}}\bigg((\alpha(r_{t}, t)\text{d}t + \beta(r_{t}, t)\text{d}W_{t})^{2} \bigg)\\
=&\hspace{1mm} \frac{\partial C}{\partial t}\text{d}t
 + \frac{\partial C}{\partial r}\bigg(\alpha(r_{t}, t)\text{d}t + \beta(r_{t}, t)\text{d}W_{t} \bigg)
 + \frac{1}{2} \frac{\partial^{2} C}{\partial r^{2}}\underbrace{\bigg(\beta(r_{t}, t)\bigg)^{2}\text{d}t}_{\text{Since (d$t^{2}=0$), (d$t$)(d$W_{t}$)$=0$, (d$W_{t}$)$^{2}=$d$t$}} \\
=&\hspace{1mm} \bigg(\frac{\partial C}{\partial t} + \alpha(r_{t}, t)\frac{\partial C}{\partial r} + \frac{1}{2} (\beta(r_{t}, t))^{2}\frac{\partial^{2} C}{\partial r^{2}} \bigg)\text{d}t + \beta(r_{t}, t)\frac{\partial C}{\partial r}\text{d}W_{t}\\
\end{align*}$$

Setting the drift term equal to $rC$, we have:
$$\frac{\partial C}{\partial t} + \alpha(r_{t}, t)\frac{\partial C}{\partial r} + \frac{1}{2} (\beta(r_{t}, t))^{2}\frac{\partial^{2} C}{\partial r^{2}}=rC$$

$$C(r, T)=F(r_{T})$$

which define the PDE and terminal condition for the pricing function $C(r, t)$.


## Problem 1b


Now supposing that the risk-neutral dynamics of $r$ are given by a Vasicek model:

$$\text{d}r_{t} = \kappa(\sigma-r_{t})\text{d}t + \sigma\text{d}W_{t},$$

i.e. $\alpha(r_{t}, t)=\kappa(\sigma-r_{t})$ and $\beta(r_{t}, t)=\sigma,$ with:
>- $\theta=0.05,$ which can be interpreted as the long-term mean of the short rate to which the short rate reverts
>- $\kappa=3,$ which can be interpreted as the speed of mean reversion
>- $\sigma=0.03,$ the constant volatility for the random component of the short rate

Then we can use the PDE derived in Problem 1(a) to price the zero-coupon bond with maturity $T=5$ and face value $F=1$:
$$\frac{\partial C}{\partial t} + \kappa\left(\sigma-r_{t}\right)\frac{\partial C}{\partial r} + \frac{1}{2} (\sigma)^{2}\frac{\partial^{2} C}{\partial r^{2}}=rC$$

Taking a `standard central-difference` explicit finite difference scheme, this can be discretised as:
$$\begin{align*}
\frac{\partial C}{\partial t} \approx &\ \frac{C^{j}_{n+1} - C^{j}_{n}}{\Delta t} \\ 
\frac{\partial C}{\partial r} \approx &\ \frac{C^{j+1}_{n+1} - C^{j-1}_{n+1}}{2\Delta r} \\ 
\frac{\partial^{2} C}{\partial r^{2}} \approx &\ \frac{C^{j+1}_{n+1} - 2C^{j}_{n+1} + C^{j-1}_{n+1}}{(\Delta r)^{2}}, \\ 
\end{align*}$$

with $C\approx C^{j}_{n}$ for the RHS of the PDE (this does not impact convergence). The discretized PDE is:
$$\frac{C^{j}_{n+1} - C^{j}_{n}}{\Delta t} + \kappa\left(\theta-r_{j}\right)\frac{C^{j+1}_{n+1} - C^{j-1}_{n+1}}{2\Delta r} + \frac{1}{2} \sigma^{2}\frac{C^{j+1}_{n+1} - 2C^{j}_{n+1} + C^{j-1}_{n+1}}{(\Delta r)^{2}}=rC^{j}_{n}$$

Rearranging for $C^{j}_{n}$, we have:
$$C^{j}_{n}=\frac{1}{1+r\Delta t}\bigg(q_{u} C^{j+1}_{n+1} + q_{m} C^{j}_{n+1} + q_{d} C^{j-1}_{n+1}\bigg),$$

where the `standard central-difference` weights are:
$$\begin{align*}
q_{u} = &\ \frac{1}{2}\left[\frac{\sigma^{2}\Delta t}{(\Delta r)^{2}} + \frac{\kappa(\theta-r_{j})\Delta t}{\Delta r}\right] \\ 
q_{m} = &\ 1 - \frac{\sigma^{2}\Delta t}{(\Delta r)^{2}} \\ 
q_{d} = &\ \frac{1}{2}\left[\frac{\sigma^{2}\Delta t}{(\Delta r)^{2}} - \frac{\kappa(\theta-r_{j})\Delta t}{\Delta r}\right] \\ 
\end{align*}$$

We will use these equations in the pricing engine for `standard central-difference`.


## Problem 1c


Now taking an explicit `upwind approximation` instead of the standard central-difference, we make an adjustment to the discretization of $\frac{\partial C}{\partial r}$ to:
$$\begin{equation}
  \frac{\partial C}{\partial r} \approx
    \begin{cases}
      \frac{C^{j+1}_{n+1} - C^{j}_{n+1}}{\Delta r} & \text{for $r_{j}$ such that }\kappa(\theta-r_{j})\geq 0\\
      \frac{C^{j}_{n+1} - C^{j-1}_{n+1}}{\Delta r} & \text{for $r_{j}$ such that }\kappa(\theta-r_{j})\lt 0\\
    \end{cases}       
\end{equation}$$

We can separate this into two cases.
- `Case 1:` For $r_{j}$ such that $\kappa(\theta-r_{j})\geq 0,$ the discretized PDE is:
$$\frac{C^{j}_{n+1} - C^{j}_{n}}{\Delta t} + \kappa\left(\theta-r_{j}\right)\frac{C^{j+1}_{n+1} - C^{j}_{n+1}}{\Delta r} + \frac{1}{2} \sigma^{2}\frac{C^{j+1}_{n+1} - 2C^{j}_{n+1} + C^{j-1}_{n+1}}{(\Delta r)^{2}}=rC^{j}_{n}$$

Rearranging for $C^{j}_{n}$, we have:
$$C^{j}_{n}=\frac{1}{1+r\Delta t}\bigg(q_{u} C^{j+1}_{n+1} + q_{m} C^{j}_{n+1} + q_{d} C^{j-1}_{n+1}\bigg),$$

where the `upwind approximation` weights for `Case 1` are:
$$\begin{align*}
q_{u} = &\ \frac{\kappa\left(\theta-r_{j}\right)\Delta t}{\Delta r} + \frac{1}{2}\frac{\sigma^{2}\Delta t}{(\Delta r)^{2}} \\ 
q_{m} = &\ 1 - \frac{\kappa\left(\theta-r_{j}\right)\Delta t}{\Delta r} - \frac{\sigma^{2}\Delta t}{(\Delta r)^{2}} \\ 
q_{d} = &\ \frac{1}{2}\frac{\sigma^{2}\Delta t}{(\Delta r)^{2}}\\ 
\end{align*}$$

- `Case 2:` For $r_{j}$ such that $\kappa(\theta-r_{j})\lt 0,$ the discretized PDE is:
$$\frac{C^{j}_{n+1} - C^{j}_{n}}{\Delta t} + \kappa\left(\theta-r_{j}\right)\frac{C^{j}_{n+1} - C^{j-1}_{n+1}}{\Delta r} + \frac{1}{2} \sigma^{2}\frac{C^{j+1}_{n+1} - 2C^{j}_{n+1} + C^{j-1}_{n+1}}{(\Delta r)^{2}}=rC^{j}_{n}$$

Rearranging for $C^{j}_{n}$, we have:
$$C^{j}_{n}=\frac{1}{1+r\Delta t}\bigg(q_{u} C^{j+1}_{n+1} + q_{m} C^{j}_{n+1} + q_{d} C^{j-1}_{n+1}\bigg),$$

where the `upwind approximation` weights for `Case 2` are:
$$\begin{align*}
q_{u} = &\ \frac{1}{2}\frac{\sigma^{2}\Delta t}{(\Delta r)^{2}} \\ 
q_{m} = &\ 1 + \frac{\kappa\left(\theta-r_{j}\right)\Delta t}{\Delta r} - \frac{\sigma^{2}\Delta t}{(\Delta r)^{2}} \\ 
q_{d} = &\ -\frac{\kappa\left(\theta-r_{j}\right)\Delta t}{\Delta r} + \frac{1}{2}\frac{\sigma^{2}\Delta t}{(\Delta r)^{2}}\\ 
\end{align*}$$

We will use these equations in the pricing engine for `upwind approximation`.

In (b) and (c), to approximate the PDE's $rC$ term, use the values of $r$ and $C$ at node $(n, j)$. (As we said in class, node $(n + 1, j)$ would also be a natural choice, but let's choose $n$ instead of $n + 1$). At the grid's upper and lower boundaries $r_{max}$ and $r_{min}$, impose $\forall t<T$ the "linearity" boundary conditions:
$$\begin{align*}
C\left(r_{max}, t\right) = &\ 2C\left(r_{max} - \Delta r, t\right) - C\left(r_{max} - 2\Delta r, t\right) \\ 
C\left(r_{min}, t\right) = &\ 2C\left(r_{min} + \Delta r, t\right) - C\left(r_{min} + 2\Delta r, t\right) \\ 
\end{align*}$$

>- This technique can help in some situations where it is not obvious what boundary conditions to use.
>- Thus, in each column of the grid, first solve for $C$ in the interior nodes; then deal with the top and bottom nodes.


In [3]:
%%time
####================================================
### Model Dynamics
class Vasicek:

    def __init__(self, kappa: float, theta: float, sigma: float) -> None:
        self.kappa = kappa
        self.theta = theta
        self.sigma = sigma


####================================================
### Contract
class Bond:

    def __init__(self, T: float) -> None:
        self.T = T



CPU times: total: 0 ns
Wall time: 0 ns


In [4]:
%%time
####================================================
### Pricing Engine
class FDexplicitEngine:

    def __init__(self, rMax: float, rMin: float, deltar: float, deltat: float, useUpwind: bool) -> None:
        self.rMax = rMax
        self.rMin = rMin
        self.deltar = deltar
        self.deltat = deltat
        self.useUpwind = useUpwind

    def price_bond_vasicek(self, contract: Bond, dynamics: Vasicek) -> Tuple[np.ndarray, np.ndarray]:
        ### Returns array of all initial short rates,
        ### and the corresponding array of zero-coupon T-maturity bond prices

        T = contract.T
        N = round(T / self.deltat)
        if abs(N - T / self.deltat) > 1e-12:
            raise ValueError('Bad delta t')
        
        ### I'm making the FIRST indices of the array correspond to HIGH levels of r
        r = np.arange(self.rMax, self.rMin - self.deltar/2, -self.deltar)   
        bondprice = np.ones(np.size(r))
        
        ### Set static parameters (make the code look cleaner)
        a = ((dynamics.sigma**2) * self.deltat) / (self.deltar**2)
        dtr = self.deltat / self.deltar

        if self.useUpwind:
            ### P1c: Apply explicit upwind approximation
            
            ### We need to implement the correct upwind approximation so we will make a mask for the index
            alpha_mask1 = (dynamics.kappa * (dynamics.theta - r) >= 0)
            alpha_mask2 = (dynamics.kappa * (dynamics.theta - r) < 0)
            
            ### Make weights into arrays to apply mask
            qu, qm, qd = np.ones(np.size(r)), np.ones(np.size(r)), np.ones(np.size(r))
            
            ### Up-State
            qu[alpha_mask1] = (dynamics.kappa * (dynamics.theta - r[alpha_mask1]) * dtr) + ((1/2) * a)
            qu[alpha_mask2] = (1/2) * a
            
            ### Mid-State
            qm[alpha_mask1] = 1 - (dynamics.kappa * (dynamics.theta - r[alpha_mask1]) * dtr) - a
            qm[alpha_mask2] = 1 + (dynamics.kappa * (dynamics.theta - r[alpha_mask2]) * dtr) - a
            
            ### Down-State
            qd[alpha_mask1] = (1/2) * a
            qd[alpha_mask2] = (-dynamics.kappa * (dynamics.theta - r[alpha_mask2]) * dtr) + ((1/2) * a)
            
        else:
            ### P1b: Apply standard central-difference equation
            b = dynamics.kappa * (dynamics.theta - r) * dtr
            qu = (1/2) * (a + b)
            qd = (1/2) * (a - b)
            qm = (1 - a) * np.ones(np.size(r))
        
        ### Compute the bond price (Induct backwards from N-1 to 0, apply bounday condition to each step)
        tn = np.arange(N-1, -1, -1)*self.deltat
        for t in tn:
            ### Note: np.roll(), the parent array is first flattened, 
            ###                  then operation is performed (shift each element by x position), 
            ###                  then the original shape is restored.
            ### Up:   C(j+1) = np.roll(bondprice, 1)
            ### Mid:  C(j+0) = np.roll(bondprice, 0) = bondprice
            ### Down: C(j-1) = np.roll(bondprice, -1)
            ### Solving Interior Nodes ###
            df_FD = 1 / (1 + r * self.deltat)
            bondprice = df_FD * (qu * np.roll(bondprice, 1) + 
                                 qm * np.roll(bondprice, 0) + 
                                 qd * np.roll(bondprice, -1))
            
            ### Solving Top and Bottom Nodes ###
            ### It is not obvious in this case, what boundary conditions to use at the top and bottom
            ### so let us assume "linearity" boundary conditions
            bondprice[0] = 2*bondprice[1] - bondprice[2]     ### C(rMax, t) = 2C(rMax-dr, t) - C(rMax-2dr, t)
            bondprice[-1] = 2*bondprice[-2] - bondprice[-3]  ### C(rMin, t) = 2C(rMin+dr, t) - C(rMin+2dr, t)

        return (r, bondprice)



CPU times: total: 0 ns
Wall time: 0 ns


In [5]:
%%time
####================================================
### Set Model Dynamics and Contract
hw3dynamics = Vasicek(kappa=3, theta=0.05, sigma=0.03)
hw3contract = Bond(T=5) 


CPU times: total: 0 ns
Wall time: 0 ns


In [6]:
%%time
####================================================
### Calculating time-0 bond prices for various r0:
for useUpwind in [False, True]:
    hw3FD = FDexplicitEngine(rMax=0.35, rMin=-0.25, deltar=0.01, deltat=0.01, useUpwind=useUpwind)
    (r, bondprice) = hw3FD.price_bond_vasicek(hw3contract, hw3dynamics)
    np.set_printoptions(precision=4, suppress=True, formatter={'float_kind':'{:f}'.format})
    displayrows = (r < 0.15 + hw3FD.deltar / 2) & (r > 0.0 - hw3FD.deltar / 2)
    
    ### Store into dataframe
    if not useUpwind:
        bondprices = pd.DataFrame(np.stack((r, bondprice), axis=1)[displayrows]).set_index(0)
    else:
        bondprices.columns = ['Standard Explicit FD']
        bondprices['Upwind Explicit FD'] = bondprice[displayrows]



CPU times: total: 46.9 ms
Wall time: 52 ms


In [7]:
%%time
####================================================
### Format dataframe and show results
bondprices.index = bondprices.index.format(formatter=lambda x: f"{x:.2f}")
bondprices.index.name = 'r0'
bondprices.style.format(formatter=lambda x: f"{x:.4f}")


CPU times: total: 719 ms
Wall time: 1.09 s


## Problem 1d


> Now let us do some comparison of the central-difference and upwind schemes.
> Suppose $f: \mathbb{R}\rightarrow\mathbb{R}$ is smooth in some open neighborhood of $x$. Show that as $h \rightarrow 0$, 
$$ \bigg\vert \frac{f(x+h) - f(x)}{h}-f{'}(x) \bigg\vert = \mathcal{O}(h) \ \ \ \bigg\vert \frac{f(x+h) - f(x-h)}{2h}-f{'}(x) \bigg\vert = \mathcal{O}(h^{2})$$
using Taylor's theorem. The $\mathcal{O}(h)$ means "some function bounded by a constant times $h$, near $h = 0$." Likewise, $\mathcal{O}(h^{2})$ means "some function bounded by a constant times $h^{2}$, near $h = 0$." Different instances of "$\mathcal{O}$" may mean different functions. The "constants" may depend on $x$ but not $h$.

Suppose $f:\mathbb{R}\rightarrow\mathbb{R}$ is smooth in some open neighborhood of $x$. From Taylor's series expansion, we have, as $h\rightarrow 0$:
$$f(x \pm h) \approx f(x) \pm hf^{'}(x) + \frac{h^{2}}{2}f^{''}(x) \pm \frac{h^{3}}{6}f^{'''}(x) + \cdots$$

For the one-sided derivative:
$$\begin{align*}
\bigg|\frac{f(x+h)-f(x)}{h}-f^{'}(x)\bigg| & = \bigg|\frac{f(x) + hf^{'}(x) + \mathcal{O}(h^{2})-f(x)}{h}-f^{'}(x)\bigg| \\
& = \bigg|f^{'}(x) + \mathcal{O}(h)-f^{'}(x)\bigg| \\
& = \mathcal{O}(h), \text{ as required} \\
\end{align*}$$

Similarly, for the centralized two-sided derivative:
$$\begin{align*}
\bigg|\frac{f(x+h)-f(x-h)}{2h}-f^{'}(x)\bigg| & = \bigg|\frac{\bigg(f(x) + hf^{'}(x) + h^{2}\frac{f^{''}(x)}{2} + \mathcal{O}(h^{3})\bigg)-\bigg(f(x) - hf^{'}(x) + h^{2}\frac{f^{''}(x)}{2} + \mathcal{O}(h^{3})\bigg)}{2h}-f^{'}(x)\bigg| \\
& = \bigg|\frac{2hf^{'}(x)+\mathcal{O}(h^{3})}{2h}-f^{'}(x)\bigg| \\
& = \bigg|f^{'}(x) + \mathcal{O}(h^{2})-f^{'}(x)\bigg| \\
& = \mathcal{O}(h^{2}), \text{ as required} \\
\end{align*}$$

>- The one-sided derivative has a convergence that is first order in $h$, or $\mathcal{O}(h)$
>- The centralized two-sided derivative has a convergence that is second order in $h$, or $\mathcal{O}(h^{2})$
>- Thus, the centralized two-sided derivative converges faster than the one-sided derivative


## Problem 1e

In [8]:
%%time
####================================================
### Using the results from the upwind explicit finite difference pricer from P1c
bondprices.loc[['0.10']]


CPU times: total: 0 ns
Wall time: 3 ms


Standard Explicit FD  Upwind Explicit FD
r0                                            
0.10            3296.59241            0.766225


>- The upwind approximation calculation is much more accurate (accurate to 3 decimal places to the true bond price, 0.7661) compared to the standard central-difference one (which gives a bond price of 3,296.59).
>- The inaccurate results from the standard explicit FD method, for high/low values of $r_0$, are due to the lack of convergence. 
>- If we print the weights ($q_{u}, q_{m}, q_{d}$) for the standard explicit FD method we get some values that are negative.
>>- In this case, we do not meet the statbility condition because we are trying to squeeze too much variance into a restricted grid spacing.
>>- If we think of this in terms of probability masses, then the sum of the absolute value of probabilities is greater than $1$, so: 
$$\lvert q_{u}\rvert + \lvert q_{m}\rvert + \lvert q_{d}\rvert > 1$$


## Problem 1f


> **Rule-of-Thumb**

Ignoring stability issues and considering only consistency (i.e. “truncation error,” also known as “local discretization error”), the upwind explicit scheme, which uses one-sided spatial differences, discretizes the PDE with <u>**less**</u> accuracy than the standard explicit scheme, which uses central spatial differences.
>- This is shown by the Taylor analysis in Problem 1d, which shows a two-sided difference converges faster (second order) than a one-sided difference (first order).

However, to actually guarantee convergence, the grid spacing must satisfy certain stability constraints. In a PDE exhibiting strong drift, we have just seen that these constraints may allow the upwind scheme <u>**greater**</u> freedom in choosing grid spacing, compared to the standard scheme.
>- If we print the weights ($q_{u}, q_{m}, q_{d}$) for the standard explicit FD method we get some values that are negative.
>>- In this case, we do not meet the statbility condition because we are trying to squeeze too much variance into a restricted grid spacing.
>>- If we think of this in terms of probability masses, then the sum of the absolute value of probabilities is greater than $1$, so: 
$$\lvert q_{u}\rvert + \lvert q_{m}\rvert + \lvert q_{d}\rvert > 1$$



## Problem 1g


> The *continuously-compounded* *yield-to-maturity* of a zero-coupon bond with time-$t$ price $P_{t}$ and nonrandom face value $P_{T}$ to be paid at maturity date $T$ is:
$$\frac{\log\left(P_{T}/P_{t}\right)}{T - t}$$
where, as always for us, $\log$ denotes natural log, and where $P_{T} = 1$ according to this problem's assumptions. One way to think of the time-$t$ yield to maturity $T$ is as the average of some type of time-$t$ expectation of the instantaneous spot rates from time $t$ to time $T$.

> Find the yield-to-maturity of a $5$-year discount bond, in the case that $r_{0} = 0.12$, and in the case that $r_{0} = 0.02$. (The "good" results from part (e) may be used here. The "bad" results should not be used, unless you want to fix them by modifying the grid spacings).
>- Why, intuitively, is the yield for $r_{0} = 0.12$ smaller than $0.12$, whereas the yield for $r_{0} = 0.02$ is greater than $0.02$?

> **Comment:** Under these short-rate dynamics, there do exist analytic pricing formulas for bonds. So we do not need finite difference methods to value the simple payoff that we have here. But the finite difference scheme can be modified to handle contracts for which exact pricing formulas do not exist.


In [9]:
%%time
####================================================
### Continuously-compounded yield-to-maturity  (YTM)
def calculate_continuous_YTM(pt: float, pT: float, T: float, t: float = 0) -> float:
    return np.log(pT/pt) / (T-t)


CPU times: total: 0 ns
Wall time: 0 ns


In [10]:
%%time
####================================================
### Get YTM from results in P1c
labelytm = 'Continuously-Compounded YTM'
bondprices[labelytm] = bondprices.apply(lambda x: calculate_continuous_YTM(x['Upwind Explicit FD'], 
                                                                           1, hw3contract.T, 0), axis=1)
bondprices.style.format(formatter=lambda x: f"{x:.4f}")


CPU times: total: 0 ns
Wall time: 1e+03 µs


In [11]:
%%time
####================================================
### Using the results from the upwind explicit finite difference pricer from P1c
bondprices.loc[['0.02', '0.12'], ['Upwind Explicit FD', 'Continuously-Compounded YTM']]


CPU times: total: 0 ns
Wall time: 3 ms


Upwind Explicit FD  Continuously-Compounded YTM
r0                                                   
0.02            0.786916                     0.047927
0.12            0.761144                     0.054587


>- For $r_{0}=0.12$ the continuously-compounded YTM is $0.0546$ which is smaller than $0.12$, whereas for $r_{0}=0.02$ the continuously-compounded YTM is $0.0479$ which is greater than $0.02$.
>- Intuitively, this can be explained by looking at the risk-neutral dynamics of $r$ as given by the Vasicek model.
>- Specifically, the drift is given by:
$$\begin{align*}
\mu_t &= \kappa(\theta - r_{t}) \\
      &= 3(0.05 - r_{t}) \\
\end{align*}$$

>- This shows that $r_{t}$ is mean-reverting, with a mean of $\theta=0.05$. 
>>- Here, volatility is a constant $\sigma$ (independent on the levels of $r_{t}$ and $t$).
>>- The long-run average $r_{t}$ is $\theta=0.05$ under risk-neutral Brownian motion. 
>>- $\kappa$ gives the speed of mean-reversion: it is positive and greater than $1$ here, and the time to maturity is long (5 years).
>>- Thus, $\kappa=3$ is sufficient for strong mean-reversion dynamics to be exhibited from time-$0$ to maturity.

>- Therefore, for $r_{0}<0.05$, we have a higher continuously-compounded YTM.
>>- Intuitively, we expect the interest rate to drift upwards in the future towards the mean level of $0.05$. 
>- Similarly, for $r_{0}>0.05$, we have a lower continuously-compounded YTM.
>>- Intuitively, we expect the interest rate to drift downwards in the future towards the mean level of $0.05$. 
>- These future interest rate movements via the mean-reverting dynamics of $r$ are captured by the backwards induction in the finite difference pricer.
>>- This is confirmed by examining the column for continuously-compounded YTM in the above dataframe.
>>- The continuously-compounded YTM for $r_{0}=\theta=0.05$ is $0.0499\approx0.05=\theta$.
>>- For $r_{0}<0.05$ the continuously-compounded YTM is greater than $0.05$.
>>- For $r_{0}>0.05$ the continuously-compounded YTM is less than $0.05$.


# Problem 2


> Suppose that a non-dividend-paying stock $S$ follows Black-Scholes dynamics, with interest rate $r$ and volatility $\sigma$.


## Problem 2a


> Given $\Delta$ where $0<\Delta<1$, and given $T>0$, solve for the strike $K$ such that a call option on $S$ with strike $K$ and expiry $T$ has time-$0$ delta equal to $\Delta$. Express your answer in terms of the inverse cdf $N^{−1}$ of the normal distribution, and any or all of $\Delta$, $S_{0}$, $r$, $\sigma$, $T$.

>- This is only valid IFF implied volatility is flat (Not a function of strike)
>- We have, from Black-Scholes, the call delta as:
$$\Delta^{C} = \frac{\partial C^{BS}}{\partial S_{t}} = \mathcal{N}(d_{1})$$
>- Then, taking the inverse, we get $d_{1}$:
$$d_{1} = \mathcal{N}^{-1}(\Delta^{C}) = \frac{log\left(\frac{S_{t}e^{r(T-t)}}{K}\right)}{\sigma \sqrt{T-t}} + \frac{1}{2} \sigma \sqrt{T-t}$$
>- From here, we can solve for the time-t strike-K:
$$K = S_{t} e^{r(T-t) - \left(\mathcal{N}^{-1}(\Delta^{C}) - \frac{1}{2} \sigma \sqrt{T-t}\right) \sigma \sqrt{T-t}}$$
>- For the time-0 strike-K:
$$K = S_{0} e^{rT - \left(\mathcal{N}^{-1}(\Delta^{C}) - \frac{1}{2} \sigma \sqrt{T}\right) \sigma \sqrt{T}}$$

>- If we consider a dividend yield, we can get the same expression for a call and put option.
>- Let $r_{Grow} = r - q$, then,
$$K_{Call} = S_{t} e^{r_{Grow}(T-t) - \left(\mathcal{N}^{-1}\left(\Delta^{C}e^{q(T-t)}\right) - \frac{1}{2} \sigma \sqrt{T-t}\right) \sigma \sqrt{T-t}}$$
$$K_{Put} = S_{t} e^{r_{Grow}(T-t) - \left(-\mathcal{N}^{-1}\left(-\Delta^{P}e^{q(T-t)}\right) - \frac{1}{2} \sigma \sqrt{T-t}\right) \sigma \sqrt{T-t}}$$


## Problem 2b


> Let $S_{0} = 300$ and $T = \frac{1}{12}$ and $\sigma = 0.4$ and $r = 0.01$. Calculate the strikes and the premiums (meaning, the prices) of a $25$-delta call and a $75$-delta call. This means that $\Delta = 0.25$ and $ \Delta = 0.75$ respectively.

> A way to get the inverse of the normal cdf in Python is: **from scipy.stats import norm** and then use **norm.ppf()**


In [12]:
%%time
####================================================
### Black-Scholes Vanilla
def vanilla_price_BS(S: float, K: float, r: float, q: float, sigma: float, T: float, t: float = 0, 
                     call_put_flag: str = 'C') -> float:
    tau = T - t
    rGrow = r - q
    d1 = np.log((S*np.exp(rGrow*tau)/K))/(sigma*np.sqrt(tau)) + (sigma*np.sqrt(tau))/2
    d2 = d1 - sigma*np.sqrt(tau)
    if call_put_flag == 'C':
        vanilla_price = S*np.exp(-q*tau)*norm.cdf(d1) - K*np.exp(-r*tau)*norm.cdf(d2)
    else:
        vanilla_price = K*np.exp(-r*tau)*norm.cdf(-d2) - S*np.exp(-q*tau)*norm.cdf(-d1)
        
    return vanilla_price

####================================================
### Flat Strike given delta
def K_from_delta(delta: float, S: float, r: float, q: float, sigma: float, T: float, t: float = 0, 
                 call_put_flag: str = 'C') -> float:
    ### This is only valid IFF implied volatility is flat (Not a function of strike)
    ### Else, use root finding method instead if their is a smile
    tau = T - t
    rGrow = r - q
    sigma_total = sigma*np.sqrt(tau)
    if call_put_flag == 'C':
        K = S*np.exp((rGrow*tau) - (norm.ppf(delta*np.exp(q*tau)) - (0.5*sigma_total))*(sigma_total))
    else:
        K = S*np.exp((rGrow*tau) - (-norm.ppf(-delta*np.exp(q*tau)) - (0.5*sigma_total))*(sigma_total))
        
    return K


CPU times: total: 0 ns
Wall time: 0 ns


In [13]:
%%time
####================================================
### Parameters
delta1 = 0.25
delta2 = 0.75
S = 300
r = 0.01
q = 0.0
sigma = 0.4
T = 1/12

### OTM: Out-of-the-Money Call Option (K_delta > S)
K25 = K_from_delta(delta1, S, r, q, sigma, T, t=0, call_put_flag='C')
call_prem25 = vanilla_price_BS(S, K25, r, q, sigma, T, t=0, call_put_flag='C')

### ITM: In-the-Money Call Option (K_delta < S)
K75 = K_from_delta(delta2, S, r, q, sigma, T, t=0, call_put_flag='C')
call_prem75 = vanilla_price_BS(S, K75, r, q, sigma, T, t=0, call_put_flag='C')

### Show premium
print('OTM: Call Delta={:.2f} has Strike-K={:.2f} and Premium=${:.2f}'.format(delta1, K25, call_prem25))
print('ITM: Call Delta={:.2f} has Strike-K={:.2f} and Premium=${:.2f}'.format(delta2, K75, call_prem75))


OTM: Call Delta=0.25 has Strike-K=326.74 and Premium=$4.88
ITM: Call Delta=0.75 has Strike-K=279.61 and Premium=$26.10
CPU times: total: 0 ns
Wall time: 971 µs


## Problem 2c


> If you spend $S_{0}$ dollars on stock, you can buy one share, and thereby obtain delta $1$.
> If you spend the same amount of $S_{0}$ dollars on, instead, an option of a given strike and expiry, then you can buy $S_{0}/C_{0}$ contracts, where $C_{0}$ is the price of $1$ contract. Each contract has delta $\Delta$, so all the contracts together have a total delta
$$\Delta \frac{S_0}{C_0}$$
> This is sometimes denoted by the Greek letter lambda, and sometimes called the "leverage" or "gearing" or "elasticity" of the option. Another way to derive/explain this lambda is that it's the *percentage* change in the option price, per *percentage* change in the underlying.

>- Calculate the lambdas of the two options in part (b). Which one gives you more leverage?


In [14]:
%%time
####================================================
### OTM: Out-of-the-Money Call Option (K_delta > S)
lambda1 = delta1*(S/call_prem25)

### ITM: In-the-Money Call Option (K_delta < S)
lambda2 = delta2*(S/call_prem75)

### Show output
print('OTM: Delta={:.2f}, S={:.2f}, Call=${:.2f}, THEN LAMBDA={:.2f}'.format(delta1, S, call_prem25, lambda1))
print('ITM: Delta={:.2f}, S={:.2f}, Call=${:.2f}, THEN LAMBDA={:.2f}'.format(delta2, S, call_prem75, lambda2))


OTM: Delta=0.25, S=300.00, Call=$4.88, THEN LAMBDA=15.36
ITM: Delta=0.75, S=300.00, Call=$26.10, THEN LAMBDA=8.62
CPU times: total: 0 ns
Wall time: 0 ns


**For call option, we have:** <br>
>- OTM: For $\Delta_{1} = 0.25, S_{0} = 300$, we get $K_{1} = 326.74$, with $\lambda_{1} = 15.36$ <br>
>- ITM: For $\Delta_{2} = 0.75, S_{0} = 300$, we get $K_{2} = 279.61$, with $\lambda_{2} = 8.62$ <br>
>- Since $\lambda_{1} > \lambda_{2}$, we have more leverage for the OTM call option.



> **Comment (no response required from you):** This is one of the reasons why, at a given strike, the $\text{OTM}$ option (call or put) is usually expected to be more liquid (more activity, narrower spreads) than the $\text{ITM}$ option (put or call) at that same strike; see the second of the three reasons below:
>- Popularity of option strategies involving $\text{OTM}$ options (for example, buying $\text{OTM}$ protective puts, and selling $\text{OTM}$ covered calls, as you implemented in the $\text{IB TWS}$ exercises Fall quarter)
>- Popularity of using options to obtain a bullish or bearish exposure to the underlying. More leverage is obtained by using $\text{OTM}$ than $\text{ITM}$ options. Thus $\text{OTM}$ options require less capital commitment to obtain a desired exposure. (To say it another way, $\text{OTM}$ options provide more exposure for a given allocation of capital).
>- $\text{ITM}$ options mainly have higher absolute deltas than $\text{OTM}$ options, so $\text{ITM}$ options are more exposed (in dollar terms, not percentage terms) to the movements of the underlying. This greater risk is a reason for market-makers to quote wider spreads for the $\text{ITM}$ option, compared to narrower spreads for the $\text{OTM}$ option.
> (Note that this $\text{OTM}$ vs. $\text{ITM}$ comparision does not compare $\text{OTM}/\text{ITM}$ vs. $\text{ATM}$ options, which are particularly active.)

> Due to the typically greater liquidity of $\text{OTM}$ options, calculations of the implied volatility surface sometimes use *only* the $\text{OTM}$ option at each strike (thus, call options at high strikes, put options at low strikes). See for instance Section 11 of Bloomberg's documentation (posted on Canvas) of how they calculate the implied vol surface.
